### PR TITLE
small `TypeVisitor` refactor

### PR DESCRIPTION
**Pull request by lcnr**
cc @LeSeulArtichaut @scottmcm 

adds `ControlFlow::map_break`

### CODE REVIEW

File: library/core/src/ops/control_flow.rs [hunk 1]
```diff
@@ -56,6 +56,20 @@ impl<B, C> ControlFlow<B, C> {
...
+    #[unstable(feature = "control_flow_enum", reason = "new API", issue = "75744")]
+    pub fn map_break<T, F>(self, f: F) -> ControlFlow<T, C>
+    where
+        F: FnOnce(B) -> T,
+    {
+        match self {
+            ControlFlow::Continue(x) => ControlFlow::Continue(x),
+            ControlFlow::Break(x) => ControlFlow::Break(f(x)),
+        }
+    }
```
Comment (LeSeulArtichaut) [inline library/core/src/ops/control_flow.rs:72]: I wonder which combinators `ControlFlow` should have, there could be different flavours of `and`s, `or`s, `map`s, `flatten`, `as_ref`...

**Review by LeSeulArtichaut — COMMENTED** [review]
:+1:

File: library/core/src/ops/control_flow.rs (hunk 1 above)
Comment (lcnr) [reply library/core/src/ops/control_flow.rs:72]: don't really know. `map_break` and `map_continue` probably make sense.

The others depending on if they are needed.

File: compiler/rustc_typeck/src/check/check.rs [hunk 2]
```diff
@@ -462,39 +462,25 @@ pub(super) fn check_opaque<'tcx>(
...
-        "check_opaque_for_inheriting_lifetimes: def_id={:?} span={:?} item={:?}",
-        def_id, span, item
-    );
-
-    #[derive(Debug)]
-    struct ProhibitOpaqueVisitor<'tcx> {
-        opaque_identity_ty: Ty<'tcx>,
-        generics: &'tcx ty::Generics,
-    }
+    debug!(?item, ?span);
```
[the remark this answers is not in the document]
Comment (lcnr) [reply compiler/rustc_typeck/src/check/check.rs:472]: part of tracing as well. it's pretty much "item={:?}, span={:?}"

File: compiler/rustc_typeck/src/check/check.rs [hunk 3]
```diff
@@ -462,39 +462,25 @@ pub(super) fn check_opaque<'tcx>(
 
 /// Checks that an opaque type does not use `Self` or `T::Foo` projections that would result
 /// in "inheriting lifetimes".
+#[instrument(skip(tcx, span))]
```
[the remark this answers is not in the document]
Comment (lcnr) [reply compiler/rustc_typeck/src/check/check.rs:465]: I think we're slowly adding it where it makes sense. Don't think there was an explicit discussion about this. cc @oli-obk

**Review by LeSeulArtichaut — APPROVED** [review]
I personally find this looks pretty good!

File: compiler/rustc_typeck/src/check/check.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jackh726) [reply compiler/rustc_typeck/src/check/check.rs:472]: >it's pretty much "item={:?}, span={:?}"

*But* you can also filter logs by these values now with this syntax

File: compiler/rustc_typeck/src/check/check.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_typeck/src/check/check.rs:465]: Yea, there's no process moving these instructions en-masse, but if you are doing any work somewhere, feel free to add them as they become helpful to you.

File: compiler/rustc_typeck/src/check/check.rs [hunk 4]
```diff
@@ -514,6 +500,26 @@ pub(super) fn check_opaque_for_inheriting_lifetimes(
...
+        opaque_identity_ty: Ty<'tcx>,
+        generics: &'tcx ty::Generics,
+    }
+
+    impl<'tcx> ty::fold::TypeVisitor<'tcx> for ProhibitOpaqueVisitor<'tcx> {
+        type BreakTy = Ty<'tcx>;
+
+        fn visit_ty(&mut self, t: Ty<'tcx>) -> ControlFlow<Self::BreakTy> {
+            debug!("check_opaque_for_inheriting_lifetimes: (visit_ty) t={:?}", t);
+            if t != self.opaque_identity_ty {
```
Comment (oli-obk) [inline compiler/rustc_typeck/src/check/check.rs:514]: Activating :paperclip: voice: Please swap the if condition and the if/else branches

**Review by oli-obk — APPROVED** [review]
r=me with a tiny nit

**Comment by LeSeulArtichaut** [comment]
Wow, I had never seen the bors queue that empty 👀

**Pull request merged**